fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/TreeLike.java

```diff
@@ -218,32 +222,66 @@ default <V extends TreeLike<V>> Optional<V> mapMaybe(@Nonnull final BiFunction<T
     @SuppressWarnings("PMD.CompareObjectsWithEquals")
     @Nonnull
     default Optional<T> replaceLeavesMaybe(@Nonnull final UnaryOperator<T> replaceOperator) {
-        return mapMaybe((t, foldedChildren) -> {
-            if (Iterables.isEmpty(foldedChildren)) {
-                return replaceOperator.apply(t);
+        return Optional.ofNullable(replace(node -> {
+            if (Iterables.isEmpty(node.getChildren())) {
+                return replaceOperator.apply(node);
             }
+            return node;
+        }));
+    }
 
-            final Iterator<? extends T> foldedChildrenIterator = foldedChildren.iterator();
-            boolean isDifferent = false;
-            for (T child : t.getChildren()) {
-                Verify.verify(foldedChildrenIterator.hasNext());
-
-                final T nextFoldedChild = foldedChildrenIterator.next();
-                if (nextFoldedChild == null) {
-                    return null;
-                }
-                // note object identity comparison is desired here as this will tell us if descendant "modified"
-                // the tree
-                if (child != nextFoldedChild) {
-                    isDifferent = true;
-                }
+    /**
+     * Replaces elements in the tree with new elements returned by the user-defined operator.
+     * It traverses the tree in pre-order fashion, and is CoW, i.e. it is guaranteed to only make a copy of a node's
+     * children iff one of the children is replaced by the replacement operator, otherwise, no allocation is made.
+     *
+     * @param replacementOperator The replacement operator defined by the user, the operator is called on every node of
+     *                            the tree, and is expected to return the node itself if no changes to be made, this is
```

Comment:
This formatting will not survive javadoc translation I think. Can you enhance, restructure?

Reply:
done.